### PR TITLE
Add `addToSetup` and helpers to `Environment`

### PR DESCRIPTION
**Pull request by soenkehahn**
This PR changes how `Environment`s set up scripts, both in the build sandboxes and outside of them. Most importantly it allows to *have* a setup for `Executable`s and `garn enter`. This would allow us to experiment with e.g. modifying the `$PATH` to add `node_modules/.bin` and similar things.

I'm not sure this is the best API, so ideas for improvements are very welcome. Also, I think the naming of everything could be better and more consistent.

### CODE REVIEW

File: ts/environment.ts [hunk 1]
```diff
@@ -24,16 +24,42 @@ import { Package, mkShellPackage } from "./package.ts";
...
+   * most common example for how the `Environment` for these sandboxes works
+   * differently is, that we have to *copy* the source files into the sandbox,
+   * to allow `Check`s and `Package`s to access them. Whereas with `Executable`s
+   * and `garn enter`, your source files are already available normally through
+   * the file system.)
+   */
+  addToSandboxSetup(snippet: string | NixExpression): Environment;
+  /**
+   * Creates a new environment based on this one that includes the specified nix
+   * packages.
```
Comment (soenkehahn) [inline ts/environment.ts:62]: Here we add `addToSetup` and `addToSandboxSetup` to the `Environment` type. I think the names of those could be better, but I hope the jsdocs explain well, what they do. Any better ideas for naming these?

File: ts/environment.ts [hunk 2]
```diff
@@ -62,11 +88,24 @@ export type Environment = {
...
 export const sandboxScript = (
   env: Environment,
   script: NixExpression,
 ): NixExpression =>
-  unlinesNixStrings([nixStrLit`mkdir -p $out`, env.sandboxSetup, script]);
+  joinNixStrings("\n", [
+    nixStrLit`mkdir -p $out`,
+    ...env.setup.map((x) => x.snippet),
+    script,
+  ]);
```
Comment (soenkehahn) [inline ts/environment.ts:108]: Here are two helpers, `commonScript` and `sandboxScript` that are used to create scripts that then run in either a build sandbox or outside of a build sandbox. Again, are there better ideas for names?

File: ts/nix.ts [hunk 3]
```diff
@@ -195,13 +195,16 @@ export function nixStrLit(
...
+  separator: string,
+  list: Array<NixExpression>,
+): NixExpression {
   return {
     [__nixExpressionTag]: null,
     type: "strLit",
     str: {
       initial: "",
-      rest: list.map((expr) => [expr, "\n"]),
+      rest: list.map((expr, i) => [expr, i < list.length - 1 ? separator : ""]),
```
Comment (alexdavid) [inline ts/nix.ts:207]: I ended up doing something very similar, except for `NixExpression`s vs `NixStrLit`s here: https://github.com/garnix-io/garn/pull/430/files#diff-fb4f9f2ab46707d603912af39f9d791c92eb585508b43f11b86e588c7c88513eR81-R86

I wonder if it makes sense to attempt to consolidate these somehow (when one of our PRs makes it in) or if the fact one operates on raw nix expressions and the other operates on string literals means they stay separate.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 ## v0.0.19 (unreleased)
 
+- Add `Environment.addToSetup` and `Environment.addToSandboxSetup`, making setups for `Environment`s more flexible.
```
Comment (alexdavid) [inline CHANGELOG.md:5]: Seeing as we discussed making these intended for internal use should we omit this changelog entry?

File: test/spec/BuildSpec.hs [hunk 5]
```diff
@@ -94,17 +94,16 @@ spec = do
...
-                echo build-content > dist/build-artifact
-              `
-            });
-            export const project = garn.mkProject({
-              description: "",
-              defaultEnvironment: env,
-            }, {})
-              .addPackage("build", "mv dist/build-artifact $out/build-artifact");
+              const env = garn.mkEnvironment()
+                .addToSandboxSetup(garn.nix.nixStrLit`
```
Comment (alexdavid) [inline test/spec/BuildSpec.hs:98]: I think these methods no longer exist on `Environment`, right?

File: ts/environment.ts [hunk 6]
```diff
@@ -62,11 +65,66 @@ export type Environment = {
...
+      },
+    ],
+  };
+}
+
+/**
+ * Add the common setup of an `Environment` to a given script. This should be
+ * used for `Executable`s and `garn enter`.
+ */
+export const commonScript = (
```
Comment (alexdavid) [inline ts/environment.ts:106]: I'm a little confused by the name here. Shouldn't this be `wrapScriptInEnv` or something?

File: ts/nix.ts (hunk 3 above)
Comment (soenkehahn) [reply ts/nix.ts]: Done in a9ba3b57.

File: CHANGELOG.md (hunk 4 above)
Comment (soenkehahn) [reply CHANGELOG.md]: Changed the changelog entry. Is that better?

**Review by alexdavid — APPROVED** [review]

**Pull request merged**